### DISCUSSION TITLE
I am getting `Internal server error` when trying to get Token

### DISCUSSION THREAD

**Discussion by SohaibAl-emara** [timdorr/tesla-api#376] [Q&A]
Hi,
I am trying to authenticate and get the token, but on the last step, i am getting this error. I was wondering if this is something caused by me trying multiples to obtain the token, or it is on the TESLA api server end. 

```JSON
{"error":"invalid_request","error_description":"Internal server error","error_uri":"https://auth.tesla.com/error/reference/f4d"}
```

**Comment by SohaibAl-emara** [comment]
Do I need MFA?

**Comment by SohaibAl-emara** [comment]
Actually, I just found out that, 
at the step where I am trying to obtain the authorisation code, my code fails and says 
```
Found. Redirecting to https://auth.tesla.com/oauth2/v3/authorize?client_id=ownerapi&code_challenge
```

**Reply by SohaibAl-emara** [reply]
Not sure what is the reason behind that

**Reply by SohaibAl-emara** [reply]
But in the redirection there is no parameter called `code=`

**Comment by magdawg** [comment]
@SohaibAl-emara I believe I'm experiencing the same issue. No `code=` in the redirection uri 
Any chance you managed to solve it?

**Reply by JonG67x** [reply]
They've just reintroduced captcha so when you make the oauth2/v3/authorize call it will return the request for a Captcha which if you don't handle properly you'll fail as there is no code. You need to work out the captcha characters and then resubmit the oauth2/v3/authorize call with all the same parameters plus the cpatch value before you get the code.

**Reply by magdawg** [reply]
@JonG67x Yup, I'm aware of the captcha. I provide `"captcha": "<value_from_captcha_svg>"` in the body of the request to the /authorize endpoint. And it returns a response with bunch of headers, including the Location header that has the redirect URI. The body of the response is something like:
```
Found. Redirecting to https://auth.tesla.com/oauth2/v3/authorize?audience=&client_id=ownerapi&code_challenge=-nZs9VDs3A_BMWn2QyjKLnU1B0SpAev57n-uozj6O6w&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fauth.tesla.com%2Fvoid%2Fcallback&response_type=code&scope=openid+email+offline_access&state=8zeITVpNtluR97mO1f8HSA
```

But I don't see a `code` in the body of the response or in the Location header.

**Reply by JonG67x** [reply]
Whats the header_code? If its 200 its asking for captcha or MFA.  I don't know how you're getting your value for the captcha.svg, it might be worth dumping that out and taking a look at the value and seeing if you agree by looking at the captcha value,  The only other tip I can give is when I was debugging it I missed one of the values on the second submission with the captcha value set and that caused it to fail. Maybe some debug statements of the various csrf, phase, transaction id, cookie etc as well as the url.

**Reply by magdawg** [reply]
It's 302. 
I'm loading the captcha in the frontend so I can see it.
Thanks for your suggestions, will keep trying.

**Reply by magdawg** [reply]
Managed to solve it by passing the Cookie in the header.
